File: web/apps/admin-web/src/api/requests.ts
import type { FreshnessContract } from "@xingmang/ui-admin";
import {
  apiClient,
  FeatureNotMountedError,
  looksLikeUnmountedRoute,
  type ApiClient,
} from "./client";
import type { ListOptions } from "./platform";

/** 请求详情（XM-0039）。数据源是生产上已在运行的外挂「请求审计系统」(reqlog)，
 *  平台只是带权限与审计的只读网关——正文永不落平台库、不缓存、默认不可导出。
 *
 *  两个端点、两个权限，分级不是「谨慎一点」而是本能力成立的前提：
 *
 *    request.read          元数据列表，回答「这个人用得多不多」
 *    request.content.read  完整正文，回答「这个人问了什么」
 *
 *  后者每读一次就在审计链上留一条 request.content.viewed（交接文档 §9.4）。 */

/** 一条请求元数据（后端 httpapi/requests.go requestSummaryItem）。
 *
 *  **没有正文字段**，这不是「前端暂时没用到」——摘要与正文的权限分级
 *  就落在这个类型边界上。 */
export interface RequestSummary {
  id: string;
  /** 来源平台：sub2api / newapi。 */
  source: string;
  occurred_at: string;
  /** 空串表示令牌没映射到用户——必须显示成「未映射」而不是空白。 */
  username: string;
  token_prefix: string;
  model: string;
  /** 本次实际路由；空串表示 reqlog 未记录。 */
  channel: string;
  upstream: string;
  /** 0 表示 reqlog 没记到状态码（连接中断）。 */
  status: number;
  duration_ms: number;
  /** null = 未记录；0 是合法观测值（缓存命中）。两者不可混同。 */
  ttfb_ms: number | null;
  tokens_in: number;
  tokens_out: number;
  tokens_cache: number;
  /** 向用户计费金额；null=未知，对象内的 0=已知零。 */
  billed_amount: RequestBilledAmount | null;
  stream: boolean;
  upstream_request_id: string;
  /** 已在后端脱敏（形如 203.0.113.x）。前端不做二次处理。 */
  client_ip: string;
}

export interface RequestBilledAmount {
  /** 十进制整数字符串，避免 JSON number 精度丢失。 */
  amount_minor: string;
  currency: string;
  scale: number;
}

export interface RequestRangeStats {
  requestCount: number;
  successCount: number;
  failureCount: number;
  averageDurationMs: number | null;
}

export interface RequestListPage {
  items: RequestSummary[];
  /** 完整过滤集统计，在游标分页前计算。 */
  stats: RequestRangeStats;
  /** 空串表示已经翻到底。不透明字符串，前端不解析、不构造。 */
  nextCursor: string;
  /** reqlog 的保留窗口。界面靠它说清「只覆盖最近 N 天」。 */
  retentionDays: number;
  /** 回答这次读取的 reqlog 实例；命中已知演示实例时挂演示横幅。 */
  dataSource: string;
  freshness: FreshnessContract;
}

export interface RequestMessage {
  role: string;
  content: string;
  truncated: boolean;
  /** 截断前的字节数。界面靠它说「只显示了 X / 共 Y」。 */
  original_bytes: number;
}

export interface RequestPayload {
  body: string;
  truncated: boolean;
  original_bytes: number;
  content_type: string;
}

export interface RequestContent {
  summary: RequestSummary;
  messages: RequestMessage[];
  /** 区分「请求体里就没有消息」与「我们没解析出来」——后者要让人去看原文。 */
  messagesParsed: boolean;
  finalReply: string;
  finalReplyTruncated: boolean;
  finalReplyBytes: number;
  rawRequest: RequestPayload;
  rawResponse: RequestPayload;
  dataSource: string;
  freshness: FreshnessContract;
}

/** 状态过滤口径。与后端逐字对应（reqlog.StatusFilter）。 */
export type RequestStatusFilter = "" | "success" | "error";

export interface RequestListFilters {
  username?: string;
  model?: string;
  status?: RequestStatusFilter;
  /** RFC3339。 */
  since?: string;
  until?: string;
}

export interface RequestListOptions extends ListOptions, RequestListFilters {
  limit?: number;
  cursor?: string;
}

/** 列表每页拉多少条。 */
export const REQUEST_PAGE_SIZE = 50;

type RawRequestSummary = Omit<RequestSummary, "billed_amount"> & {
  billed_amount?: unknown;
};

interface RawRequestStats {
  request_count: number;
  success_count: number;
  failure_count: number;
  average_duration_ms: number | null;
}

interface RawRequestPage {
  items: RawRequestSummary[] | null;
  stats: RawRequestStats;
  next_cursor?: string;
  retention_days?: number;
  data_source?: string;
  freshness: FreshnessContract;
}

interface RawRequestContent {
  summary: RawRequestSummary;
  messages: RequestMessage[] | null;
  messages_parsed: boolean;
  final_reply: string;
  final_reply_truncated: boolean;
  final_reply_bytes: number;
  raw_request: RequestPayload;
  raw_response: RequestPayload;
  data_source?: string;
  freshness: FreshnessContract;
}

/** 空串不进 URL：「不传」和「传空串」对后端不是一回事（见 client.ts）。 */
function omitEmpty(value: string | undefined): string | undefined {
  const trimmed = (value ?? "").trim();
  return trimmed === "" ? undefined : trimmed;
}

function parseBilledAmount(raw: unknown): RequestBilledAmount | null {
  if (raw === null || raw === undefined) return null;
  if (typeof raw !== "object") throw new Error("请求计费数据格式异常");
  const value = raw as Record<string, unknown>;
  const amountMinor = value.amount_minor;
  const currency = value.currency;
  const scale = value.scale;
  if (
    typeof amountMinor !== "string" ||
    !/^\d+$/.test(amountMinor) ||
    typeof currency !== "string" ||
    currency.trim() === "" ||
    typeof scale !== "number" ||
    !Number.isInteger(scale) ||
    scale < 0 ||
    scale > 18
  ) {
    throw new Error("请求计费数据格式异常");
  }
  return { amount_minor: amountMinor, currency, scale };
}

function parseRequestSummary(raw: RawRequestSummary): RequestSummary {
  return {
    id: raw.id,
    source: raw.source,
    occurred_at: raw.occurred_at,
    username: raw.username,
    token_prefix: raw.token_prefix,
    model: raw.model,
    channel: raw.channel ?? "",
    upstream: raw.upstream ?? "",
    status: raw.status,
    duration_ms: raw.duration_ms,
    ttfb_ms: raw.ttfb_ms,
    tokens_in: raw.tokens_in,
    tokens_out: raw.tokens_out,
    tokens_cache: raw.tokens_cache,
    billed_amount: parseBilledAmount(raw.billed_amount),
    stream: raw.stream,
    upstream_request_id: raw.upstream_request_id,
    client_ip: raw.client_ip,
  };
}

function parseCount(value: unknown, label: string): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) {
    throw new Error(`请求统计 ${label} 格式异常`);
  }
  return value;
}

function parseStats(raw: RawRequestStats): RequestRangeStats {
  const requestCount = parseCount(raw?.request_count, "请求数");
  const successCount = parseCount(raw?.success_count, "成功数");
  const failureCount = parseCount(raw?.failure_count, "失败数");
  const average = raw?.average_duration_ms;
  if (average !== null && (typeof average !== "number" || !Number.isSafeInteger(average) || average < 0)) {
    throw new Error("请求统计平均耗时格式异常");
  }
  if (successCount + failureCount !== requestCount) {
    throw new Error("请求统计计数不一致");
  }
  if ((requestCount === 0) !== (average === null)) {
    throw new Error("请求统计平均耗时与请求数不一致");
  }
  return { requestCount, successCount, failureCount, averageDurationMs: average };
}

/** 请求详情「未接入」的说明文案（列表、正文共用同一句话——它们是同一条链路
 *  的一体两面，理由不该分叉）。
 *
 *  `XM_REQLOG_MODE=off` 时后端整组不挂载 `/platforms/{platform}/requests*`
 *  （cmd/platform-api/reqlog.go newRequestLogService），前端据此分辨
 *  「没接」与「坏了」，不能显示成加载失败。 */
const REQUESTS_NOT_MOUNTED_DESCRIPTION =
  "请求详情在当前环境未启用（XM_REQLOG_MODE=off）。接入真实 reqlog 数据源后会自动出现，无需手动开启。";

/** 列出某平台的请求元数据（需 request.read）。 */
export async function listPlatformRequests(
  platform: string,
  options: RequestListOptions = {},
  client: ApiClient = apiClient,
): Promise<RequestListPage> {
  let body: RawRequestPage;
  try {
    body = await client.get<RawRequestPage>(
      `/api/v1/platforms/${encodeURIComponent(platform)}/requests`,
      {
        searchParams: {
          limit: String(options.limit ?? REQUEST_PAGE_SIZE),
          cursor: omitEmpty(options.cursor),
          username: omitEmpty(options.username),
          model: omitEmpty(options.model),
          status: omitEmpty(options.status),
          since: omitEmpty(options.since),
          until: omitEmpty(options.until),
        },
        ...(options.signal ? { signal: options.signal } : {}),
      },
    );
  } catch (error) {
    // XM_REQLOG_MODE=off 时这条端点整组不挂载：chi 的默认 404 没有可解析的
    // error.code，用它区分「没接」与「这次请求恰好失败了」
    if (looksLikeUnmountedRoute(error)) {
      throw new FeatureNotMountedError(error, REQUESTS_NOT_MOUNTED_DESCRIPTION);
    }
    throw error;
  }
  return {
    items: (body.items ?? []).map(parseRequestSummary),
    stats: parseStats(body.stats),
    nextCursor: body.next_cursor ?? "",
    retentionDays: body.retention_days ?? 0,
    dataSource: body.data_source ?? "",
    freshness: body.freshness,
  };
}

/** 读取单条请求的完整正文（需 request.content.read）。
 *
 *  **每次调用都会在审计链上留一条 request.content.viewed。** 所以它不该被
 *  预取、不该在鼠标悬停时触发、也不该在列表里批量调用——那会让「谁看过什么」
 *  这份清单里混进大量没人真的看过的条目。调用点只有请求详情页一个。
 *
 *  `reason` 目前可选（是否必填是产品决定，见后端 ContentInput.Reason）。 */
export async function getPlatformRequestContent(
  platform: string,
  requestId: string,
  options: ListOptions & { reason?: string } = {},
  client: ApiClient = apiClient,
): Promise<RequestContent> {
  let body: RawRequestContent;
  try {
    body = await client.get<RawRequestContent>(
      `/api/v1/platforms/${encodeURIComponent(platform)}/requests/${encodeURIComponent(requestId)}`,
      {
        searchParams: { reason: omitEmpty(options.reason) },
        ...(options.signal ? { signal: options.signal } : {}),
      },
    );
  } catch (error) {
    // 具体请求 id 不存在时后端仍会挂载路由、走 WriteError 回一个带 code 的
    // 404（见 requests_test.go）；只有整组端点没挂载才会落进这一支
    if (looksLikeUnmountedRoute(error)) {
      throw new FeatureNotMountedError(error, REQUESTS_NOT_MOUNTED_DESCRIPTION);
    }
    throw error;
  }
  return {
    summary: parseRequestSummary(body.summary),
    messages: body.messages ?? [],
    messagesParsed: body.messages_parsed,
    finalReply: body.final_reply,
    finalReplyTruncated: body.final_reply_truncated,
    finalReplyBytes: body.final_reply_bytes,
    rawRequest: body.raw_request,
    rawResponse: body.raw_response,
    dataSource: body.data_source ?? "",
    freshness: body.freshness,
  };
}

File: platform/internal/platform/httpapi/requests.go
package httpapi

import (
	"context"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/xufei5620/xingmang-platform/connectors/reqlog"
	"github.com/xufei5620/xingmang-platform/internal/platform/action"
	"github.com/xufei5620/xingmang-platform/internal/platform/principal"
	"github.com/xufei5620/xingmang-platform/internal/platform/requestlog"
)

// RequestLogQuerier 是请求详情的只读能力（*requestlog.Service 满足）。
type RequestLogQuerier interface {
	List(ctx context.Context, in requestlog.ListInput) (reqlog.RequestLogPage, error)
	Content(ctx context.Context, in requestlog.ContentInput) (reqlog.RequestLogContent, error)
}

// requestSummaryItem 是一条请求元数据的对外表示。
//
// **逐字段列出，不是 reqlog.RequestLogSummary 的直接序列化**：响应体是契约，
// 不是结构体的倒影（规格 §18.4）。这条在本端点上不只是洁癖——契约层将来若
// 为了排查方便往 Summary 上加一个带正文片段的字段，直接序列化会让它一路
// 漏到只有 request.read 权限的人眼前，而这里加字段必须是一次显式动作。
type requestSummaryItem struct {
	ID         string `json:"id"`
	Source     string `json:"source"`
	OccurredAt string `json:"occurred_at"`
	// Username 为空表示令牌没映射到用户——前端必须显示成「未映射」
	// 而不是空白（空白会被读成「这条没有用户」）。
	Username    string `json:"username"`
	TokenPrefix string `json:"token_prefix"`
	Model       string `json:"model"`
	Channel     string `json:"channel"`
	Upstream    string `json:"upstream"`
	Status      int    `json:"status"`
	DurationMS  int64  `json:"duration_ms"`
	// TTFBMS 为 null 表示未记录；0 是合法观测值（缓存命中），两者不可混同。
	TTFBMS      *int64 `json:"ttfb_ms"`
	TokensIn    int64  `json:"tokens_in"`
	TokensOut   int64  `json:"tokens_out"`
	TokensCache int64  `json:"tokens_cache"`
	// BilledAmount 是向终端用户计费金额；null=未知，对象内的 0=已知零。
	BilledAmount      *requestBilledAmount `json:"billed_amount"`
	Stream            bool                 `json:"stream"`
	UpstreamRequestID string               `json:"upstream_request_id"`
	// ClientIP 已在连接器侧脱敏（reqlog.MaskIP），这里不做二次处理——
	// 二次处理会让「脱敏在哪一层做」变成两个答案。
	ClientIP string `json:"client_ip"`
}

type requestBilledAmount struct {
	AmountMinor string `json:"amount_minor"`
	Currency    string `json:"currency"`
	Scale       int32  `json:"scale"`
}

type requestStats struct {
	RequestCount      int64  `json:"request_count"`
	SuccessCount      int64  `json:"success_count"`
	FailureCount      int64  `json:"failure_count"`
	AverageDurationMS *int64 `json:"average_duration_ms"`
}

// requestPage 是一页请求元数据。
type requestPage struct {
	Items []requestSummaryItem `json:"items"`
	Stats requestStats         `json:"stats"`
	// NextCursor 为空串表示已经翻到底。
	NextCursor string `json:"next_cursor"`
	// RetentionDays 让界面能就地说清「只覆盖最近 N 天」——
	// 「那天没有请求」与「过了保留期」在屏幕上长得一模一样，处置却完全不同。
	RetentionDays int `json:"retention_days"`
	// DataSource 是回答这次读取的 reqlog 实例（fake 时为 reqlog-fake）。
	//
	// 字段名不叫 source：item 里的 `source` 已经是**平台**（sub2api/newapi），
	// 两个同名字段指两件事，读响应的人一定会搞混。这里对齐
	// /api/v1/metrics 里 source 的**语义**（哪个实例产的这批数），
	// 前端据此挂演示数据横幅。
	DataSource string `json:"data_source"`
	// Freshness 与 /api/v1/metrics 同一个形状，前端复用同一个徽章组件。
	// 不是可选字段（规格 §9.1：禁止裸数字冒充实时完整数据）。
	Freshness freshnessBody `json:"freshness"`
}

type requestMessage struct {
	Role      string `json:"role"`
	Content   string `json:"content"`
	Truncated bool   `json:"truncated"`
	// OriginalBytes 让界面说得出「只显示了 X / 共 Y」。
	// 一个不声张的截断比不显示更危险——人会以为看到的就是全部。
	OriginalBytes int64 `json:"original_bytes"`
}

type requestPayload struct {
	Body          string `json:"body"`
	Truncated     bool   `json:"truncated"`
	OriginalBytes int64  `json:"original_bytes"`
	ContentType   string `json:"content_type"`
}

// requestContentBody 是单条请求的完整内容（高敏）。
type requestContentBody struct {
	Summary  requestSummaryItem `json:"summary"`
	Messages []requestMessage   `json:"messages"`
	// MessagesParsed 区分「请求体里就没有消息」与「我们没解析出来」——
	// 后者要让人去看原文，前者不用。
	MessagesParsed      bool           `json:"messages_parsed"`
	FinalReply          string         `json:"final_reply"`
	FinalReplyTruncated bool           `json:"final_reply_truncated"`
	FinalReplyBytes     int64          `json:"final_reply_bytes"`
	RawRequest          requestPayload `json:"raw_request"`
	RawResponse         requestPayload `json:"raw_response"`
	// DataSource 与列表页同义：fake 时为 reqlog-fake。详情页也要挂横幅——
	// 恰恰是这一页最容易被当成真实用户问答截图转发出去。
	DataSource string        `json:"data_source"`
	Freshness  freshnessBody `json:"freshness"`
}

// requestLogStalenessThresholdSeconds 是这条通道的新鲜度阈值。
//
// reqlog 是抄录器而不是采集任务：这里的「观测时刻」就是我们向它发问的时刻，
// 所以正常情况下 staleness 恒等于零、状态恒为 fresh。阈值仍要有且要小——
// 它守的不是上游的采集周期，而是**我们自己这条链路**：一个耗时超过一分钟
// 才返回的读取，屏幕上那批数字已经不能当「刚刚」用了。
const requestLogStalenessThresholdSeconds int32 = 60

// freshnessFromSnapshot 把连接器的快照折算成统一的新鲜度形状。
//
// 状态优先级与 ops.Observation.Freshness 保持一致（stale 先于 partial）：
// 两处若给出不同的优先级，同一个徽章在两个页面上会对同一种情况说两种话。
// 这里天然产生不了 failed / uninitialized——读失败根本不会走到这个函数
// （错误已经在 Service 层翻译成了 HTTP 错误），而「从未采集」对一次实时读取
// 不是一种可能的状态。
func freshnessFromSnapshot(snap reqlog.Snapshot, now time.Time) freshnessBody {
	observedAt := snap.ObservedAt.UTC()
	staleness := int64(now.Sub(observedAt).Seconds())
	if staleness < 0 {
		// 上游时钟比我们快时会出现负数。夹到 0 而不是原样返回：
		// 一个「负 3 秒前的数据」会让人以为界面坏了，而它其实只是时钟偏移
		staleness = 0
	}
	body := freshnessBody{
		StalenessSeconds: &staleness,
		ThresholdSeconds: requestLogStalenessThresholdSeconds,
		IsPartial:        snap.IsPartial,
		ObservedAt:       rfc3339Ptr(&observedAt),
		LastSuccess:      rfc3339Ptr(&observedAt),
	}
	switch {
	case staleness >= int64(requestLogStalenessThresholdSeconds):
		body.State = "stale"
	case snap.IsPartial:
		body.State = "partial"
	default:
		body.State = "fresh"
	}
	return body
}

func toSummaryItem(s reqlog.RequestLogSummary) (requestSummaryItem, error) {
	var billedAmount *requestBilledAmount
	if s.BilledAmount != nil {
		if err := s.BilledAmount.Validate(); err != nil {
			return requestSummaryItem{}, fmt.Errorf("invalid billed amount: %w", err)
		}
		billedAmount = &requestBilledAmount{
			AmountMinor: strconv.FormatInt(s.BilledAmount.AmountMinor, 10),
			Currency:    s.BilledAmount.Currency,
			Scale:       s.BilledAmount.Scale,
		}
	}
	return requestSummaryItem{
		ID:     s.ID,
		Source: s.Source,
		// 时间一律 UTC（宪法 14 条），本地化交给前端
		OccurredAt:        s.OccurredAt.UTC().Format(time.RFC3339),
		Username:          s.Username,
		TokenPrefix:       s.TokenPrefix,
		Model:             s.Model,
		Channel:           s.Channel,
		Upstream:          s.Upstream,
		Status:            s.Status,
		DurationMS:        s.DurationMS,
		TTFBMS:            s.TTFBMS,
		TokensIn:          s.TokensIn,
		TokensOut:         s.TokensOut,
		TokensCache:       s.TokensCache,
		BilledAmount:      billedAmount,
		Stream:            s.Stream,
		UpstreamRequestID: s.UpstreamRequestID,
		ClientIP:          s.ClientIP,
	}, nil
}

// parseTimeParam 解析 RFC3339 时间参数；缺省返回零值。
//
// 解析失败当成参数错（400）而不是静默忽略：静默忽略会把一次
// 「只看今天」的查询悄悄变成「看全部」，而返回的条数看起来完全正常。
func parseTimeParam(r *http.Request, name string) (time.Time, error) {
	raw := strings.TrimSpace(r.URL.Query().Get(name))
	if raw == "" {
		return time.Time{}, nil
	}
	t, err := time.Parse(time.RFC3339, raw)
	if err != nil {
		return time.Time{}, action.NewError(action.CodeInvalidParams,
			name+" 必须是 RFC3339 时间（如 2026-08-28T09:00:00Z）", err)
	}
	return t.UTC(), nil
}

func parseStatusParam(r *http.Request) (reqlog.StatusFilter, error) {
	raw := reqlog.StatusFilter(strings.TrimSpace(r.URL.Query().Get("status")))
	switch raw {
	case reqlog.StatusAny, reqlog.StatusSuccess, reqlog.StatusError:
		return raw, nil
	default:
		return "", action.NewError(action.CodeInvalidParams,
			"status 只接受 success / error 或留空", nil)
	}
}

func parseLimitParam(r *http.Request) (int, error) {
	raw := strings.TrimSpace(r.URL.Query().Get("limit"))
	if raw == "" {
		return 0, nil // 0 交给连接器用默认页大小
	}
	v, err := strconv.Atoi(raw)
	if err != nil || v < 0 {
		return 0, action.NewError(action.CodeInvalidParams, "limit 必须是非负整数", err)
	}
	return v, nil
}

// ListPlatformRequestsHandler 列出某平台的请求**元数据**（不含正文）。
//
// 环境范围：**不接受 environment 查询参数**，一律用 Principal 的环境——
// 与审计事件同一档处理，理由也相同：这批数据逐条可还原一个人的使用轨迹，
// 多一个可写的入参就多一个将来被放宽成「跨环境看板」的口子（规格 §20.5）。
//
// 说得更具体些：reqlog 是一个部署在特定环境里的抄录器，本进程只连得到自己
// 那一个。所谓「跨环境读取」在这条链路上连物理可能性都没有，因此把
// environment 做成参数只会制造一个看起来能用、实际永远无效的旋钮。
//
// 权限（requestlog.ScopeRead）由路由上的 RequireScope 判定，不在此处重复。
func ListPlatformRequestsHandler(q RequestLogQuerier) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if _, ok := principal.FromContext(r.Context()); !ok {
			WriteError(w, r, action.NewError(action.CodePermissionDenied, "缺少身份", nil))
			return
		}
		status, err := parseStatusParam(r)
		if err != nil {
			WriteError(w, r, err)
			return
		}
		since, err := parseTimeParam(r, "since")
		if err != nil {
			WriteError(w, r, err)
			return
		}
		until, err := parseTimeParam(r, "until")
		if err != nil {
			WriteError(w, r, err)
			return
		}
		limit, err := parseLimitParam(r)
		if err != nil {
			WriteError(w, r, err)
			return
		}

		page, err := q.List(r.Context(), requestlog.ListInput{
			Platform: chi.URLParam(r, "platform"),
			Username: strings.TrimSpace(r.URL.Query().Get("username")),
			Model:    strings.TrimSpace(r.URL.Query().Get("model")),
			Status:   status,
			Since:    since,
			Until:    until,
			Limit:    limit,
			Cursor:   strings.TrimSpace(r.URL.Query().Get("cursor")),
		})
		if err != nil {
			WriteError(w, r, err)
			return
		}

		items := make([]requestSummaryItem, 0, len(page.Items))
		for _, s := range page.Items {
			item, err := toSummaryItem(s)
			if err != nil {
				WriteError(w, r, action.NewError(action.CodeInternal,
					"请求计费数据格式异常", err))
				return
			}
			items = append(items, item)
		}
		WriteJSON(w, http.StatusOK, requestPage{
			Items: items,
			Stats: requestStats{
				RequestCount:      page.Stats.RequestCount,
				SuccessCount:      page.Stats.SuccessCount,
				FailureCount:      page.Stats.FailureCount,
				AverageDurationMS: page.Stats.AverageDurationMS,
			},
			NextCursor:    page.NextCursor,
			RetentionDays: page.RetentionDays,
			DataSource:    page.Instance,
			Freshness:     freshnessFromSnapshot(page.Snapshot, time.Now().UTC()),
		})
	}
}

// GetPlatformRequestContentHandler 读取单条请求的完整内容（高敏）。
//
// 三条与列表端点不同的纪律：
//
//	权限更高一档 → requestlog.ScopeContentRead（路由上判）；
//	每次落审计   → Service.Content 负责，写不进去就不返回内容；
//	不缓存       → /api/v1 整组已有 NoStore（XM-0031），本端点因此天然继承。
//	              这不是"顺带"：正文一旦进了浏览器磁盘缓存或中间代理，
//	              §9.4「默认禁止导出」就成了一句只在按钮层面成立的话。
//
// `reason` 是可选查询参数，原样写进审计事件的 reason 列。做成可选而不是必填，
// 是因为「每看一条都要打字」是一条会改变日常操作手感的产品决定
// （见 requestlog.ContentInput.Reason 的说明）。
func GetPlatformRequestContentHandler(q RequestLogQuerier) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		p, ok := principal.FromContext(r.Context())
		if !ok {
			WriteError(w, r, action.NewError(action.CodePermissionDenied, "缺少身份", nil))
			return
		}

		content, err := q.Content(r.Context(), requestlog.ContentInput{
			Principal: p,
			Platform:  chi.URLParam(r, "platform"),
			ID:        chi.URLParam(r, "requestID"),
			Reason:    strings.TrimSpace(r.URL.Query().Get("reason")),
			// 请求 ID 取自 RequestID 中间件，与访问日志、错误响应里的是同一个：
			// 审计事件因此能和那两处对上（交接文档 §9.4 要求记录 request_id）
			RequestID: RequestIDFrom(r.Context()),
		})
		if err != nil {
			WriteError(w, r, err)
			return
		}

		messages := make([]requestMessage, 0, len(content.Messages))
		for _, m := range content.Messages {
			messages = append(messages, requestMessage{
				Role: m.Role, Content: m.Content,
				Truncated: m.Truncated, OriginalBytes: m.OriginalBytes,
			})
		}
		summary, err := toSummaryItem(content.Summary)
		if err != nil {
			WriteError(w, r, action.NewError(action.CodeInternal,
				"请求计费数据格式异常", err))
			return
		}
		WriteJSON(w, http.StatusOK, requestContentBody{
			Summary:             summary,
			Messages:            messages,
			MessagesParsed:      content.MessagesParsed,
			FinalReply:          content.FinalReply,
			FinalReplyTruncated: content.FinalReplyTruncated,
			FinalReplyBytes:     content.FinalReplyBytes,
			RawRequest:          toPayload(content.RawRequest),
			RawResponse:         toPayload(content.RawResponse),
			DataSource:          content.Instance,
			Freshness:           freshnessFromSnapshot(content.Snapshot, time.Now().UTC()),
		})
	}
}

func toPayload(p reqlog.RawPayload) requestPayload {
	return requestPayload{
		Body: p.Body, Truncated: p.Truncated,
		OriginalBytes: p.OriginalBytes, ContentType: p.ContentType,
	}
}
